### PR TITLE
[feature] Pull docs from OpenWISP modules to create a unified documentation Website #107

### LINKED ISSUE #107: [docs] Comprehensive/Unified Documentation Website
We need a comprehensive documentation which describes all the modules.

I want to avoid having the documentation of each module contained in this repo because that would make it harder to maintain: if we keep the documentation of each module in the module itself, we will be able to always request contributors to document their work because it will be easy for them.

We could use a system of git submodules, each OpenWISP module will be a submodule in this repo and we'll have some scripts which will glue everything together and generate the sphinx documentation.

The documentation could be divided as follows:

- user documentation (quickstart, guides, explanation of features)
- developer (information for developers and contributors)
- general (project goals, values, press, etc)

The user and developer sections will then have one sub-directory for each module, or something similar.

Each module could have a structure like the following:

```
# user:
/openwisp-controller/docs/user
/openwisp-controller/docs/user/feature1
/openwisp-controller/docs/user/feature2
# docs:
/openwisp-controller/docs/developer
/openwisp-controller/docs/developer/setup
/openwisp-controller/docs/developer/extending
```

The `Makefile` will take each module `user` and `developer` directory and copy the files as follows:

```
cp /openwisp-controller/docs/user/* /openwisp2-docs/user/openwisp-controller/
cp /openwisp-controller/docs/developer/* /openwisp2-docs/developer/openwisp-controller/
```

### PR DESCRIPTION
**Pull request by pandafy**
Closes #107

### CODE REVIEW

**Review by nemesifier — COMMENTED** [review]
We can publish this on GC in the `__new__` subdirectory.

File: _static/index.html [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta http-equiv="refresh" content="0; URL='/22.05/index.html'" />
```
Comment (nemesifier) [inline _static/index.html:4]: This looks fragile: what happens when we release a new stable version?
Is it possible to have an alias for the latest stable version?

File: config.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      - name: openwisp-monitoring
+        branch: reorder-docs
+      - name: openwisp-notifications
+        branch: reorder-docs
+      - name: openwisp-network-topology
+        branch: reorder-docs
+      - name: openwisp-users
+        branch: reorder-docs
+      - name: openwisp-utils
+        branch: reorder-docs
```
Comment (nemesifier) [inline config.yml:21]: Can we change the implementation so that the order of this list is respected and the index here reflects this?

Screenshot from 2024-04-29 21-26-20

Without this change I am not sure we have a way to decide the order of these documents, which is problematic.

File: index.rst [hunk 3]
```diff
@@ -3,40 +3,65 @@ OpenWISP Documentation
...
+      developer/hacktoberfest
+
+.. ifconfig:: version != '22.05'
+
+   .. toctree::
+      :maxdepth: 1
+      :glob:
+
+      user/quickstart
+      modules/*/docs/*
```
Comment (nemesifier) [inline index.rst:51]: maybe we need to name the folders with numbers to order them as we want? Would that impact the URLs?

File: _static/index.html (hunk 1 above)
Comment (pandafy) [reply _static/index.html:4]: I couldn't find a way to render an HTML file with context from sphinx configuration.

File: index.rst (hunk 3 above)
Comment (pandafy) [reply index.rst:51]: Yes, that would change the URLs.

File: config.yml (hunk 2 above)
Comment (pandafy) [reply config.yml:21]: Instead of using `modules/*/docs/*` in the TOC tree. We can specify the path of each module to get the desired order.

**Review by nemesifier — CHANGES_REQUESTED** [review]
Please rebase on the latest master and see my comments below.

File: pulling_modules.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        os.environ['OPENWISP2_VERSION'] = version_name
+        subprocess.run(
+            ['sphinx-build', '-b', 'html', '.', f'_build/{version_name}'], check=True
+        )
+        subprocess.run(
+            ['sphinx-build', '-b', 'pdf', '.', f'_build/{version_name}'], check=True
+        )
+        subprocess.run(
+            ['sphinx-build', '-b', 'epub', '.', f'_build/{version_name}'], check=True
+        )
```
[the remark this answers is not in the document]
Comment (pandafy) [reply pulling_modules.py:69]: @nemesifier are you suggesting to remove the Python script from here and do everything in the Makefile?

**Review by nemesifier — CHANGES_REQUESTED** [review]
I am happy of the progress we are making here! :clap: :fire:  :rocket: 

Here's the summary of the improvements we have discussed.

### Stable alias

We need to generate a symbolic link "stable", pointing to the latest available stable version.

We can generate it only if the following conditions are true:

- The link doesn't exist
- Or it exists and point to a different version that what we have computed to be the latest.

To compute the latest version, instead of having to specify it manually, let's determine that automatically.
The stable version is the highest non dev version, we can compute it by getting all available versions, and getting the higher value (eg: use `max(version_list)`).

### YAML: avoid repeating modules

Let's avoid having to repeat the available modules in each version, eg:

```yaml 
---

modules:
  - name: openwisp-controller
    dir_name: controller
  - name: openwisp-monitoring
    dir_name: monitoring
  - name: openwisp-firmware-upgrader
    dir_name: firwmare-upgrader
  - name: openwisp-notifications
    dir_name: notifications
  - name: openwisp-network-topology
    dir_name: network-topology
  - name: openwisp-users
    dir_name: users
  - name: openwisp-utils
    dir_name: utils
versions:
  # just for example purposes
  # you can see the branch is missing
  # because it's implicit, we don't need to specify it
  - name: "24.06"
  - name: "22.05"
  - name: dev
    branch: reorder-docs
```

### Menu is shown twice

The menu is shown twice, we need to have 1 menu only.

### Separate current 22.05 docs from dev

We need to use the 22.05 branch for stable docs, so that we can do any change we need in the master branch without creating conflicts or missing pages with the old stable docs.

### Simplify URL structure

With the current structure we would end up with an URL like the following:

`https://openwisp.io/docs/stable/modules/openwisp-controller/docs/user/templates.html`.

The goal is to get more friendly URLs like:

`https://openwisp.io/docs/stable/controller/user/templates.html`.

Here's a possible solution I thought, which as a bonus also allows to avoid building files like `README.rst`, `CONTRIBUTING.rst`, `CHANGES.rst` which we should not build.

Every time we clone/checkout, we copy or link the docs/ directory to the root of the docs repository, renaminig it to the value of the `dir_name` attribute in the YAML.

We have to instruct sphinx to not build anything in the `modules directory`, the following seems to work for me:

```diff
diff --git a/conf.py b/conf.py
index e711b80..dd19d8c 100644
--- a/conf.py
+++ b/conf.py
@@ -85,7 +85,7 @@ language = 'en'
 # List of patterns, relative to source directory, that match files and
 # directories to ignore when looking for source files.
 # This patterns also effect to html_static_path and html_extra_path
-exclude_patterns = ['README.rst', '_build', 'Thumbs.db', '.DS_Store']
+exclude_patterns = ['README.rst', '_build', 'Thumbs.db', '.DS_Store', 'modules/*']
 
 # The reST default role (used for this markup: `text`) to use for all
 # documents.
```

Once the build step is completed, we can remove the copies/link.

We need to add those dirs in the .gitignore so we won't commit anything by mistake.

### Allow building only a specific version

The previous step will make building only a specific version harder, due to the copying/linking renaming and deleting of the docs directory each time.

Can you add an option to allow do that through the Makefile?

### Include more modules

Include the following repos:

- IPAM
- openwisp-config (let's use openwrt-config-agent as `dir_name`)
- openwrt-openwisp-monitoring (let's use openwrt-monitoring-agent as `dir_name`)
- wifi-login-pages
- ansible-openwisp2
- docker-openwisp

File: build.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+OUTPUT_FORMATS = ['pdf', 'epub', 'html']
+
+
+def clone_or_update_repo(module_name, branch):
+    """
+    Clone or update a repository based on the module name and branch provided.
+    If the repository already exists, update it. Otherwise, clone the repository.
+    """
+    repo_url = f'https://github.com/openwisp/{module_name}.git'
```
Comment (nemesifier) [inline build.py:24]: I was thinking here, since our repos are open source and public, I think we can use the git protocol here instead of SSH, isnt't it?

This will make it easier for us to work on the entire set of documents from one place and push changes to all the modules more easily. Now I found myself having to use `git remote set-url origin ...` a lot before being able to push.

**Review by nemesifier — CHANGES_REQUESTED** [review]

**Review by nemesifier — CHANGES_REQUESTED** [review]
While reviewing and improving the docs of OpenWISP Controller, I found quite a lot of broken links.
Our build here is supposed to be able to detect broken links, but I think the changes of this PR are making the `run-qa-checks` script ineffective, so please double check that, ensure that broken links are detected.

**Review by nemesifier — COMMENTED** [review]
Another thing we will need to verify is that PDF opens correctly.
A few days ago I couldn't open the generated PDF, now the error went away.
However, it's better to have an automatic verification mechanism.

**Comment by pandafy** [comment]
**TODOS**

- [ ] Generate symbolic link for stable version of each module 
- [ ] Support passing module details from environment variable 
- [ ] Build documentation for 22.05 separately. This will automatically fix duplicated menu entries. 
- [ ] Allow building only specific versions using make file
- [ ] Fix the QA check for broken links (This has been fixed, I have commented the code for now to allow build to pass)
- [ ] Verify PDFs open correctly

File: Makefile [hunk 6]
```diff
@@ -52,7 +52,7 @@ clean:
 
 .PHONY: build
 build:
-	./build.py $(formats)
+	python3 build.py $(formats)
```
Comment (nemesifier) [inline Makefile:55]: let's not do this, let the env figure out the right python version to use by adding in the beginning of `build.py` the following shebang:

```
#!/usr/bin/env python
```

File: build.py [hunk 7]
```diff
@@ -16,12 +16,17 @@
...
 
 
-def clone_or_update_repo(module_name, branch):
+def clone_or_update_repo(module_name, branch, is_production=False):
     """
     Clone or update a repository based on the module name and branch provided.
     If the repository already exists, update it. Otherwise, clone the repository.
     """
-    repo_url = f'https://github.com/openwisp/{module_name}.git'
+    if is_production:
```
Comment (nemesifier) [inline build.py:45]: I am not convinced this approach is comfortable.
This is the standard case right? So I think we should default to this case, without the need of having to set any env variable.
The other case of using the git protocol and SSH, is the special case we need to be aware of.
So we could require setting the env var for that case as "DEV=1". We need to document this in the README.

**Review by nemesifier — CHANGES_REQUESTED** [review]

**Review by nemesifier — CHANGES_REQUESTED** [review]
Got this exception while trying to build the HTML documentation for `dev` with `make build_html VERSION=dev`, this is the stacktrace:

```
Traceback (most recent call last):
  File "/openwisp/openwisp2-docs/./build.py", line 381, in <module>
    main()
  File "/openwisp/openwisp2-docs/./build.py", line 335, in main
    clone_or_update_repo(
  File "/openwisp/openwisp2-docs/./build.py", line 272, in clone_or_update_repo
    shutil.copytree(os.path.join(clone_path, 'docs'), dir_name)
  File "/usr/lib/python3.10/shutil.py", line 559, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/lib/python3.10/shutil.py", line 457, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "/usr/lib/python3.10/os.py", line 225, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'controller'
make[1]: *** [Makefile:56: build] Error 1
make[1]: Leaving directory '/openwisp/openwisp2-docs'
make: *** [Makefile:60: build_html] Error 2
```
Can you please make sure the code is resilient to these kind of failures?

File: README.rst [hunk 8]
```diff
@@ -17,18 +17,151 @@ Requirements: Python >= 3.9.
...
+
+By default, the OpenWISP modules are cloned over HTTPS. This may pose
+a hurdle if you wish to make changes to the cloned modules and push
+them to the remote URL. To use SSH remotes, you can set the environment
+variable ``DEV=1``. This will instruct the build to clone the modules
+using SSH instead of HTTPS. For example:
+
+.. code-block:: shell
+
+    DEV=1 make build
```
Comment (nemesifier) [inline README.rst:151]: Since we have a `dev` version, the choice of `DEV` here makes it confusing.
Why not just use `SSH=1`?

File: README.rst [hunk 9]
```diff
@@ -17,18 +17,151 @@ Requirements: Python >= 3.9.
...
+
+Building for production
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to build the documentation for the production environment,
+i.e. for deploying on `https://openwisp.io/docs`, set the environment
+variable ``PRODUCTION=1``. This will generate the documentation in the
+following formats:
```
Comment (nemesifier) [inline README.rst:159]: what formats?

File: README.rst [hunk 10]
```diff
@@ -17,18 +17,151 @@ Requirements: Python >= 3.9.
...
+a hurdle if you wish to make changes to the cloned modules and push
+them to the remote URL. To use SSH remotes, you can set the environment
+variable ``DEV=1``. This will instruct the build to clone the modules
+using SSH instead of HTTPS. For example:
+
+.. code-block:: shell
+
+    DEV=1 make build
+
+Building for production
```
Comment (nemesifier) [inline README.rst:153]: Why is this necessary? I don't understand what this does.

File: README.rst [hunk 11]
```diff
@@ -17,18 +17,151 @@ Requirements: Python >= 3.9.
...
+
+    version=<openwisp-version>,repository=<repo-owner>/<repo-name>,branch=<branch-name>
+
+E.g. if you want to build the documentation for the ``dev`` version, but want
+to use the ``feature`` branch of openwisp-controller of your fork, then
+the command will be:
+
+.. code-block:: shell
+
+    make build MODULES="version=dev:repository=<your-username>/openwisp-controller:branch=feature"
```
Comment (nemesifier) [inline README.rst:131]: If we already have the version flag, why allow to specify the version here too?

Seems redundant to me. 

Why not just:

```
make build VERSION=dev MODULES="repository=<your-username>/openwisp-controller:branch=feature"
```

Can you allow something like this?

```
make build VERSION=dev MODULES="openwisp-users,openwisp-controller"
```

In this case, if nothing is specified, we can assume this is the list of module we want to build, so use the defaults provided in the config YAML but build only these modules instead of all of them.

**Review by nemesifier — CHANGES_REQUESTED** [review]
Another issue I encountered.

Command: `make build_html MODULES="version=dev:repository=openwisp/openwisp-users:branch=reorder-docs`.

Stacktrace:

```
Traceback (most recent call last):
  File "/openwisp/openwisp2-docs/./build.py", line 381, in <module>
    main()
  File "/openwisp/openwisp2-docs/./build.py", line 335, in main
    clone_or_update_repo(
  File "/openwisp/openwisp2-docs/./build.py", line 272, in clone_or_update_repo
    shutil.copytree(os.path.join(clone_path, 'docs'), dir_name)
  File "/usr/lib/python3.10/shutil.py", line 559, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/lib/python3.10/shutil.py", line 457, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "/usr/lib/python3.10/os.py", line 225, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'monitoring'
make[1]: *** [Makefile:56: build] Error 1
make[1]: Leaving directory '/openwisp/openwisp2-docs'
make: *** [Makefile:60: build_html] Error 2
```

Anything wrong with the command I used?

Is it possible to specify only which module to build and which version? Without having to specify the repository or branch?

See also my comments below.

File: README.rst (hunk 11 above)
Comment (pandafy) [reply README.rst:131]: > Can you allow something like this?
> 
> ```
> make build VERSION=dev MODULES="openwisp-users,openwisp-controller"
> ```
> 
> In this case, if nothing is specified, we can assume this is the list of module we want to build, so use the defaults provided in the config YAML but build only these modules instead of all of them.

How will this help us in using the script programmatically?

If you want to do this for quick builds, you can temporarily comment out the modules from config.yml

Comment (pandafy) [reply README.rst:131]: > If we already have the version flag, why allow to specify the version here too?
> 
> Seems redundant to me.
> 
> Why not just:
> 
> ```
> make build VERSION=dev MODULES="repository=<your-username>/openwisp-controller:branch=feature"
> ```

The `VERSION` argument determines which version to build. The first `version` key in the `MODULES` determine which version to add this module. 

E.g. you can have something like this 

```
make build MODULES="version=dev:repository=<your-username>/openwisp-controller:branch=feature,version=22.05:repository=<your-username>/openwisp-controller:branch=22.05"
```

File: README.rst [hunk 12]
```diff
@@ -17,18 +17,151 @@ Requirements: Python >= 3.9.
 
 2. Clone this repository using the following command::
 
-    git clone https://github.com/<your-username>/openwisp2-docs.git
```
Comment (nemesifier) [inline README.rst:20]: Let's rename the title of this document to "OpenWISP Documentation".
I renamed the repo to openwisp-docs (from openwisp2-docs).

File: build.py [hunk 13]
```diff
@@ -268,6 +268,8 @@ def clone_or_update_repo(name, branch, dir_name, owner='openwisp'):
         )
     # If the module contains a doc directory, copy it to the dir_name in the root.
     # Otherwise, copy the entire directory.
+    if os.path.exists(dir_name):
```
Comment (nemesifier) [inline build.py:271]: Can you try with symbolic links instead of copying the entire dir to see if it works? If it does it should make the process faster.

File: run-qa-checks [hunk 14]
```diff
@@ -16,10 +16,26 @@ openwisp-qa-check \
 
 echo ''
 echo 'Checking documentation build status'
```
Comment (nemesifier) [inline run-qa-checks:18]:
```suggestion
# In order to validate the documentation, we need to build it.
echo 'Build the documentation, any warning will fail the build'
```

File: build.py [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+    with open('config.yml') as f:
+        config = yaml.safe_load(f)
+
+    build_versions = get_build_versions(config['versions'], args.version)
+    stable_version = get_stable_version(build_versions)
+    docs_root = ''
+    html_base_url = ''
+    build_dir = '_build'
+    if os.environ.get('PRODUCTION', False):
```
Comment (nemesifier) [inline build.py:310]: we may remove this if

File: build.py [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+        docs_root = '/docs/__new__'
```
Comment (nemesifier) [inline build.py:311]: Pseudo code:
```suggestion
        docs_root = '$(DOCS_ROOT)'
```

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -39,19 +39,25 @@ jobs:
         run: sudo npm install -g stylelint
 
       - name: QA checks
-        run: ./run-qa-checks
+        run: PRODUCTION=1 ./run-qa-checks
```
Comment (nemesifier) [inline .github/workflows/ci.yml:42]:
```suggestion
        # In this repository, the run-qa-checks script also builds the actual output
        run: DOCS_ROOT='/docs/__new__' ./run-qa-checks
```

File: README.rst (hunk 10 above)
Comment (nemesifier) [reply README.rst:153]: Maybe we can remove this.

File: README.rst (hunk 11 above)
Comment (nemesifier) [reply README.rst:131]: Ok let's leave this for now.

**Review by nemesifier — CHANGES_REQUESTED** [review]

File: version_switcher/builders.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
+from sphinx.builders.dummy import DummyBuilder
+import json
+from .utils import load_versions_map, VERSION_MAP_FILE
+
+
+class OpenwispDummyBuilder(DummyBuilder):
+    name = 'ow_dummy'
```
Comment (nemesifier) [inline version_switcher/builders.py:15]: can you add a docstring explaining what this class does?
Can we find a better name for this? Like `VersionSwitcherIndexBuilder` or anything similar?

File: Makefile [hunk 19]
```diff
@@ -240,3 +257,7 @@ dummy:
 	$(SPHINXBUILD) -b dummy $(ALLSPHINXOPTS) $(BUILDDIR)/dummy
 	@echo
 	@echo "Build finished. Dummy builder generates no files."
+
+.PHONY: ow_dummy
+ow_dummy:
```
Comment (nemesifier) [inline Makefile:262]: same here, can we find a more descriptive name?

File: Makefile [hunk 20]
```diff
@@ -240,3 +257,7 @@ dummy:
...
+	$(SPHINXBUILD) -b ow_dummy $(ALLSPHINXOPTS) $(BUILDDIR)/ow_dummy
```
Comment (nemesifier) [inline Makefile:263]: same for `$(BUILDDIR)/ow_dummy`

**Review by nemesifier — CHANGES_REQUESTED** [review]

**Review by nemesifier — COMMENTED** [review]
Now that https://github.com/openwisp/openwisp-docs/pull/194 is merged, please double check if we need to update anything here. I guess we don't here but we need to update the ansible-openwisp2 repo.

File: run-qa-checks [hunk 21]
```diff
@@ -16,10 +16,26 @@ openwisp-qa-check \
...
+# cd _build/
+# python -m http.server 8001 &> /dev/null & pid=$!
+# sleep 4
+# BASE_URL="http://localhost:8001"
+# if [ -n "${PRODUCTION:-}" ]; then
+#     BASE_URL="$BASE_URL/docs/__new__"
+# fi
+# pylinkvalidate.py "$BASE_URL/22.05/" \
+#     -w 4 -m process \
+#     --ignore "$BASE_URL/22.05/controller,$BASE_URL/22.05/monitoring,$BASE_URL/22.05/firmware-upgrader,$BASE_URL/22.05/network-topology,$BASE_URL/22.05/notifications,$BASE_URL/22.05/users,$BASE_URL/22.05/utils"
```
Comment (nemesifier) [inline run-qa-checks:32]: can we re-enable this now?

File: user/django-settings.rst [hunk 22]
```diff
@@ -68,14 +78,27 @@ add any python instruction to define and manipulate settings via the
...
+
+- :doc:`OpenWISP Controller Settings </controller/user/settings>`
+- :doc:`OpenWISP Monitoring Settings </monitoring/user/settings>`
+- :doc:`OpenWISP Firmware Upgrader Settings
+  </firmware-upgrader/user/settings>`
+- :doc:`OpenWISP Network Topology Settings
+  </network-topology/user/settings>`
+- :doc:`OpenWISP Users Settings </users/user/settings>`
+- :doc:`OpenWISP Notifications Settings </notifications/user/settings>`
+- :doc:`OpenWISP Utils Settings </utils/user/settings>`
```
Comment (nemesifier) [inline user/django-settings.rst:104]: Isn't ipam missing here?

**Review by nemesifier — CHANGES_REQUESTED** [review]

File: user/django-settings.rst (hunk 22 above)
Comment (pandafy) [reply user/django-settings.rst:104]: IPAM does not have any Django settings

File: run-qa-checks (hunk 21 above)
Comment (pandafy) [reply run-qa-checks]: Yes, I have enabled it.

File: tutorials/wpa-enterprise-eap-ttls-pap.rst [hunk 23]
```diff
@@ -217,7 +217,7 @@ Paste the following JSON in the ``Raw JSON Editing`` field.
     :target: ../_images/config-variable-mac-json.png
 
 You can refer the :doc:`Configuration Variables
-</user/configuration-variables>` section of this documentation for more
+</controller/user/variables>` section of this documentation for more
```
Comment (nemesifier) [inline tutorials/wpa-enterprise-eap-ttls-pap.rst:220]: this is not the only link that needs to be updated here, please update the link to the page "Configuring FreeRADIUS for WPA Enterprise (EAP-TTLS-PAP)" in the ansible module. 

There's many instances of other links that need to be updated in these pages too, please revise by looking for "<https://github.com/openwisp".

**Review by nemesifier — CHANGES_REQUESTED** [review]

**Review by nemesifier — APPROVED** [review]
